EngineLayer/MetaMorpheusEngine.cs

```diff
@@ -172,6 +221,50 @@ public static List<MatchedFragmentIon> MatchFragmentIons(Ms2ScanWithSpecificMass
             return matchedFragmentIons;
         }
 
+        public static List<MatchedFragmentIon> MatchFragmentIonsOfAllCharges(Ms2ScanWithSpecificMass scan, List<Product> theoreticalProducts, CommonParameters commonParameters)
+        {
+            var matchedFragmentIons = new List<MatchedFragmentIon>();
+            var Ions = new List<string>();
+           
+            // if the spectrum has no peaks
+            if (scan.ExperimentalFragments != null && !scan.ExperimentalFragments.Any())
+            {
+                return matchedFragmentIons;
+            }
+
+            // search for ions in the spectrum
+            //foreach (Product product in theoreticalProducts)
+            for (int i = 0; i < theoreticalProducts.Count; i++)
+            {
+                var product = theoreticalProducts[i];
+                // unknown fragment mass; this only happens rarely for sequences with unknown amino acids
+                if (double.IsNaN(product.NeutralMass))
+                {
+                    continue;
+                }
+
+                var minMass = commonParameters.ProductMassTolerance.GetMinimumValue(product.NeutralMass);
+                var maxMass = commonParameters.ProductMassTolerance.GetMaximumValue(product.NeutralMass);
+                var closestExperimentalMassList = scan.GetClosestExperimentalIsotopicEnvelopeList(minMass, maxMass);
+                if (closestExperimentalMassList != null)
+                {
+                    foreach (var x in closestExperimentalMassList)
+
+                    {
+                        String Ion = product.ProductType.ToString() + product.FragmentNumber + "^" + x.Charge + "-" + product.NeutralLoss;
```

Comment:
lower case

Reply:
This method risks being super slow because of the long string lookups. Particularly in top-down where there are many deconvoluted experimental masses and many high scoring PrSMs. Might want to consider a way to refactor.

Reply:
Can also use the $"{string1}{string2}" syntax. https://docs.microsoft.com/en-us/dotnet/csharp/how-to/concatenate-multiple-strings#string-interpolation

Reply:
Will a Tuple<Product , int > be better?
Codes are like: 
var ions = new List<Tuple<Product, int>>();
foreach (var x in closestExperimentalMassList)
 {
     Tuple<Product, int> ion = new Tuple<Product, int>(product, x.Charge);
     ....... 
}